File: server_integration_tests/kotlin/kotlin_client/client-app/src/test/kotlin/dev/graphlink/kotlinclienttest/SubscriptionsTest.kt
package dev.graphlink.kotlinclienttest

import dev.graphlink.kotlinclient.generated.client.GraphLinkClient
import dev.graphlink.kotlinclient.generated.inputs.CreateArticleInput
import dev.graphlink.kotlinclient.generated.inputs.UpdateArticleInput
import kotlinx.coroutines.async
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.time.Duration.Companion.seconds

class SubscriptionsTest {

    private lateinit var client: GraphLinkClient

    @BeforeTest
    fun setUp() {
        client = Fixtures.newClient()
    }

    @Test
    fun articleCreated_emitsWhenArticleIsCreated() = runBlocking {
        val eventDeferred = async { client.subscriptions.articleCreated().first() }

        // give the subscription a moment to register before triggering the mutation
        delay(300)

        val created = client.mutations.createArticle(CreateArticleInput(title = "Subscribed Post", authorId = "1"))

        val event = withTimeout(10.seconds) { eventDeferred.await() }
        assertEquals(created.createArticle.id, event.articleCreated.id)
        assertEquals("Subscribed Post", event.articleCreated.title)
    }

    @Test
    fun articleUpdated_emitsForMatchingArticle() = runBlocking {
        val created = client.mutations.createArticle(CreateArticleInput(title = "Will Update", authorId = "2"))

        val eventDeferred = async { client.subscriptions.articleUpdated(created.createArticle.id).first() }

        delay(300)

        val updated = client.mutations.updateArticle(
            UpdateArticleInput(id = created.createArticle.id, title = "Updated via subscription")
        )

        val event = withTimeout(10.seconds) { eventDeferred.await() }
        assertEquals(updated.updateArticle.id, event.articleUpdated.id)
        assertEquals("Updated via subscription", event.articleUpdated.title)
    }
}

File: server_integration_tests/clients/swift_client/Tests/GraphLinkGeneratedTests/SubscriptionsTest.swift
import XCTest
@testable import GraphLinkGenerated

final class SubscriptionsTest: XCTestCase {

    func testArticleCreated_emitsWhenAnArticleIsCreated() async throws {
        let client = Fixtures.newClient()

        let stream = client.subscriptions.articleCreated(fieldArgs: ArticleCreatedFieldArgs(latestArticlesLimit: 2))
        let eventTask = Task {
            try await firstValue(from: stream, timeoutSeconds: 10)
        }

        try await Task.sleep(nanoseconds: 300_000_000)

        let created = try await client.mutations.createArticle(
            input: CreateArticleInput(title: "Subscribed Post", authorId: "1"),
            fieldArgs: CreateArticleFieldArgs(latestArticlesLimit: 2)
        )

        let event = try await eventTask.value
        XCTAssertEqual(event.articleCreated.id, created.createArticle.id)
        XCTAssertEqual(event.articleCreated.title, "Subscribed Post")
    }

    func testArticleUpdated_emitsWhenTheMatchingArticleIsUpdated() async throws {
        let client = Fixtures.newClient()

        let created = try await client.mutations.createArticle(
            input: CreateArticleInput(title: "Will Update", authorId: "2"),
            fieldArgs: CreateArticleFieldArgs(latestArticlesLimit: 2)
        )

        let stream = client.subscriptions.articleUpdated(
            id: created.createArticle.id,
            fieldArgs: ArticleUpdatedFieldArgs(latestArticlesLimit: 2)
        )
        let eventTask = Task {
            try await firstValue(from: stream, timeoutSeconds: 10)
        }

        try await Task.sleep(nanoseconds: 300_000_000)

        let updated = try await client.mutations.updateArticle(
            input: UpdateArticleInput(id: created.createArticle.id, title: "Updated via subscription"),
            fieldArgs: UpdateArticleFieldArgs(latestArticlesLimit: 2)
        )

        let event = try await eventTask.value
        XCTAssertEqual(event.articleUpdated.id, updated.updateArticle.id)
        XCTAssertEqual(event.articleUpdated.title, "Updated via subscription")
    }
}
